### PR TITLE
[ticket/12618] Allow extension author to use SSL for version-check.

### PR DESCRIPTION
**Pull request by kilianr**
Fixes Ticket 12618 (https://tracker.phpbb.com/browse/PHPBB3-12618)

In order to be able to use SSL with fsockopen, one has to append 'ssl://' to the host. Thus, the class file_downloader needed to be changed.
'ssl://' will always be appended now, if the port is 443.
I have checked, that this does not influence any other uses of that class (there are none).

### CODE REVIEW

**Comment by bantu** [comment]
Unfortunately the way in which SSL/TLS is used here lacks security in the sense that certificates are not validated at all.

**Comment by kilianr** [comment]
@bantu Ok, maybe I'm missing the point here. So far, it has been impossible to connect via https to get the information for the version check. This PR would solve that. True, for older versions of php, fsockopen doesn't care about certificates (it does starting with PHP 5.6 by the way).
So in order to make https connections at least possible it would still be nice to merge this PR, but if you have other requirements, its ok for me to close it. However, in this case, it would be nice to state those requirements in the ticket (e.g. what are trusted root-certs, to you want to pin some certs, do you want to consider DNSSec, maybe even DANE, ....) because right now, they are unclear.
My personal opinion would be, that if you just validate that the certificate is for the domain, you gain nothing. So you need to establish a chain of trust, which brings us back to root-certs or DANE.

**Comment by Nicofuma** [comment]
imo, the best solution would be to use `Guzzle` in the `file_downloader` instead of using the sockets manually.

ButI don't think it can be done in 3.1.x

File: phpBB/phpbb/file_downloader.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ public function get($host, $directory, $filename, $port = 80, $timeout = 6)
 		$this->error_number = 0;
 		$this->error_string = '';
 
-		if ($socket = @fsockopen($host, $port, $this->error_number, $this->error_string, $timeout))
+		if ($socket = @fsockopen(($port == 443 ? 'ssl://' : '') . $host, $port, $this->error_number, $this->error_string, $timeout))
```
Comment (bantu) [inline phpBB/phpbb/file_downloader.php]: This should certainly be tls://

**Comment by bantu** [comment]
@kilianr Maybe this patch isn't all too bad. My problem is mostly the lack of certificate checking and lack of authentication as a result. Maybe we should just decide that this is a not an issue for version information.

**Comment by Elsensee** [comment]
> Maybe we should just decide that this is a not an issue for version information.

Yes. Also, as @kilianr already said, this is fixed with PHP 5.6.

**Comment by Nicofuma** [comment]
I think we can say it's not an issue for version check.

:+1: 

@bantu you confirm it's good for you?

File: phpBB/phpbb/file_downloader.php (hunk 1 above)
Comment (marc1706) [inline phpBB/phpbb/file_downloader.php]: Yeah, I'd go for ```tls://``` here

Comment (kilianr) [inline phpBB/phpbb/file_downloader.php]: Changed it to tls:// in last recent commit.

**Comment by Elsensee** [comment]
Thank you. :) :+1:

**Pull request merged**